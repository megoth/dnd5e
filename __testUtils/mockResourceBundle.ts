import { mockSolidDatasetFrom } from "@inrupt/solid-client";
import { ResourceBundleModel } from "../src/models/resourceBundle";
import mockFluentBundle from "./mockFluentBundle";
import { createSWRResponse } from "./mockSWR";
import { appVocabURL } from "./mockAppIndexDataset";

export const errorsIndexURL = "https://example.com/errors.ttl";
export const faqIndexURL = "https://example.com/faq.ttl";
export const translationsIndexURL = "https://example.com/translations.ttl";
export const localizedIndexURL = "https://example.com/translations.en-US.ttl";

export default function mockResourceBundle(
  overrides: Partial<ResourceBundleModel> = {}
): ResourceBundleModel {
  return {
    appVocabURL,
    bundleNames: ["global"],
    currentLanguage: "en-US",
    errorsIndexURL: { global: errorsIndexURL },
    errorsIndexSWR: {
      global: createSWRResponse(mockSolidDatasetFrom(errorsIndexURL)),
    },
    faqIndexURL: { global: faqIndexURL },
    faqIndexSWR: {
      global: createSWRResponse(mockSolidDatasetFrom(faqIndexURL)),
    },
    fluentBundles: {
      global: [mockFluentBundle({}, "en-US")],
    },
    localizedIndexURL: { global: localizedIndexURL },
    localizedIndexSWR: {
      global: createSWRResponse(mockSolidDatasetFrom(faqIndexURL)),
    },
    translationsIndexURL: { global: translationsIndexURL },
    translationsIndexSWR: {
      global: createSWRResponse(mockSolidDatasetFrom(translationsIndexURL)),
    },
    ...overrides,
  };
}
