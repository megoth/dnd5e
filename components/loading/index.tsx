import React from "react";

export const TESTID_LOADING = "loading";

export default function Loading() {
  // Do not try to translate anything here - it's used when loading translations as well
  return <div data-testid={TESTID_LOADING}>Loading...</div>;
}
