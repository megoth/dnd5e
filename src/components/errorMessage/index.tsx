import React from "react";
import NestedError from "nested-error-stacks";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";

export default function ErrorMessage({ error }: Props) {
  // const { message: errorURL } = error;
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        {/*{translationURL ? (*/}
        {/*  <Localized id={getTranslationId(translationURL)}>*/}
        {/*    {errorURL}*/}
        {/*  </Localized>*/}
        {/*) : (*/}
        {error.message}
        {/*)}*/}
      </h1>
    </div>
  );
}
