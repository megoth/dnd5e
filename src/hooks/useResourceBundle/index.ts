import { useContext } from "react";
import { ResourceBundleContext } from "../../contexts/resourceBundle";

export default function useResourceBundle() {
  return useContext(ResourceBundleContext);
}
