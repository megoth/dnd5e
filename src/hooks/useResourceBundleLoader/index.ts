import useSWR from "swr";
import { useEffect, useState } from "react";
import NestedError from "nested-error-stacks";
import useAppIndex from "../useAppIndex";
import { getLanguages } from "../../models/language";
import useDataset from "../useDataset";
import useFluentBundles from "../useFluentBundles";
import { ResourceBundleModel } from "../../models/resourceBundle";
import { currentLanguage } from "../../models/translation";
import useLanguages from "../useLanguages";

export const swrConfig = {
  errorRetryCount: 0,
};

export default function useResourceBundleLoader(
  bundleName: string,
  appIndexURL,
  appVocabURL
) {
  const languages = useLanguages();
  const currentLocales = getLanguages(currentLanguage, languages);
  const { data: appIndex, error: appIndexError } = useAppIndex(
    currentLocales,
    appIndexURL,
    appVocabURL
  );
  const [bundleModel, setBundleModel] = useState<ResourceBundleModel>({
    appVocabURL,
    bundleNames: [],
    currentLanguage,
    errorsIndexSWR: {},
    errorsIndexURL: {},
    faqIndexSWR: {},
    faqIndexURL: {},
    fluentBundles: {},
    localizedIndexSWR: {},
    localizedIndexURL: {},
    translationsIndexSWR: {},
    translationsIndexURL: {},
  });
  const [errorsIndexURL, setErrorsIndexURL] = useState<string>();
  const errorsIndexSWR = useDataset(errorsIndexURL, [bundleModel], swrConfig);
  const [faqIndexURL, setFaqIndexURL] = useState<string>();
  const faqIndexSWR = useDataset(faqIndexURL, [bundleModel], swrConfig);
  const [translationsIndexURL, setTranslationsIndexURL] = useState<string>();
  const translationsIndexSWR = useDataset(
    translationsIndexURL,
    [bundleModel],
    swrConfig
  );
  const [localizedIndexURL, setLocalizedIndexURL] = useState<string>();
  const localizedIndexSWR = useDataset(
    localizedIndexURL,
    [bundleModel],
    swrConfig
  );
  const fluentBundles = useFluentBundles(localizedIndexSWR, currentLocales);
  useEffect(() => {
    if (bundleModel.bundleNames.indexOf(bundleName) !== -1 || appIndexError) {
      return;
    }
    setBundleModel({
      ...bundleModel,
      bundleNames: bundleModel.bundleNames.concat(bundleName),
    });
  }, [appIndexError, bundleModel, bundleModel.bundleNames, bundleName]);
  return useSWR(
    [
      bundleModel.bundleNames,
      bundleName,
      appIndex,
      errorsIndexSWR,
      faqIndexSWR,
      translationsIndexSWR,
      localizedIndexSWR,
      fluentBundles,
    ],
    () => {
      if (appIndexError) {
        throw new NestedError("Unable to load resource bundle", appIndexError);
      }
      if (!appIndex) {
        return null;
      }
      const bundle = appIndex.resourceBundleAll.find(
        ({ label }) => label === bundleName
      );
      if (!bundle) {
        return null;
      }
      if (errorsIndexURL !== bundle.errorsIndexURL) {
        setErrorsIndexURL(bundle.errorsIndexURL);
      }
      if (faqIndexURL !== bundle.faqIndexURL) {
        setFaqIndexURL(bundle.faqIndexURL);
      }
      if (translationsIndexURL !== bundle.translationsIndexURL) {
        setTranslationsIndexURL(bundle.translationsIndexURL);
      }
      if (localizedIndexURL !== bundle.localizedIndexURL) {
        setLocalizedIndexURL(bundle.localizedIndexURL);
      }
      return fluentBundles
        ? {
            appVocabURL,
            bundleNames: bundleModel.bundleNames,
            currentLanguage: bundleModel.currentLanguage,
            errorsIndexSWR: {
              ...bundleModel.errorsIndexSWR,
              [bundleName]: errorsIndexSWR,
            },
            errorsIndexURL: {
              ...bundleModel.errorsIndexURL,
              [bundleName]: errorsIndexURL,
            },
            faqIndexSWR: {
              ...bundleModel.faqIndexSWR,
              [bundleName]: faqIndexSWR,
            },
            faqIndexURL: {
              ...bundleModel.faqIndexURL,
              [bundleName]: faqIndexURL,
            },
            fluentBundles: {
              ...bundleModel.fluentBundles,
              [bundleName]: fluentBundles,
            },
            localizedIndexSWR: {
              ...bundleModel.localizedIndexSWR,
              [bundleName]: localizedIndexSWR,
            },
            localizedIndexURL: {
              ...bundleModel.localizedIndexURL,
              [bundleName]: localizedIndexURL,
            },
            translationsIndexSWR: {
              ...bundleModel.translationsIndexSWR,
              [bundleName]: translationsIndexSWR,
            },
            translationsIndexURL: {
              ...bundleModel.translationsIndexURL,
              [bundleName]: translationsIndexURL,
            },
          }
        : null;
    },
    swrConfig
  );
}
