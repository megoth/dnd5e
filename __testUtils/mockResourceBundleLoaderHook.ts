import { responseInterface } from "swr";
import mockResourceBundle from "./mockResourceBundle";
import { ResourceBundleModel } from "../src/models/resourceBundle";

export default function mockResourceBundleLoaderHook(
  mock: jest.Mock,
  swrResponse: Partial<responseInterface<ResourceBundleModel, any>> = {
    data: mockResourceBundle(),
  }
) {
  return mock.mockReturnValue(swrResponse);
}
