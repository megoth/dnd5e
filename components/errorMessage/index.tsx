import React from "react";
import { Localized } from "@fluent/react";
import NestedError from "nested-error-stacks";
import useResourceBundle from "../../src/hooks/useResourceBundle";
import { getTranslationId } from "../../src/models/translation";
import { getErrorTranslationURL } from "../../src/models/error";

interface Props {
  error: NestedError;
}

export const TESTID_ERROR = "error";
export const TESTID_ERROR_TITLE = "error-title";

export default function ErrorMessage({ error }: Props) {
  const { resourceBundle } = useResourceBundle();
  const { message: errorURL } = error;
  const translationURL = resourceBundle
    ? getErrorTranslationURL(errorURL, resourceBundle)
    : null;
  return (
    <div data-testid={TESTID_ERROR}>
      <h1 data-testid={TESTID_ERROR_TITLE}>
        {translationURL ? (
          <Localized id={getTranslationId(translationURL)}>
            {errorURL}
          </Localized>
        ) : (
          errorURL
        )}
      </h1>
    </div>
  );
}
