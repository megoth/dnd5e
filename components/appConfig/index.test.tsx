import React from "react";
import { render } from "@testing-library/react";
import useResourceBundleLoader from "../../src/hooks/useResourceBundleLoader";
import AppConfig from "./index";
import mockResourceBundleLoaderHook from "../../__testUtils/mockResourceBundleLoaderHook";
import { TESTID_ERROR } from "../errorMessage";
import useResourceBundle from "../../src/hooks/useResourceBundle";
import mockResourceBundle from "../../__testUtils/mockResourceBundle";
import { TESTID_LOADING } from "../loading";
import {
  appIndexURL,
  appVocabURL,
} from "../../__testUtils/mockAppIndexDataset";

jest.mock("../../src/hooks/useResourceBundleLoader");
const mockedResourceBundleLoaderHook = useResourceBundleLoader as jest.Mock;

jest.mock("../../src/hooks/useResourceBundle");
const mockedResourceBundleHook = useResourceBundle as jest.Mock;

const resourceBundle = mockResourceBundle();

describe("AppConfig", () => {
  beforeEach(() =>
    mockedResourceBundleHook.mockReturnValue({ resourceBundle })
  );

  it("loads resources and renders", () => {
    mockResourceBundleLoaderHook(mockedResourceBundleLoaderHook, {
      data: resourceBundle,
    });
    const { asFragment } = render(
      <AppConfig appIndexURL={appIndexURL} appVocabURL={appVocabURL}>
        test
      </AppConfig>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(mockedResourceBundleLoaderHook).toHaveBeenCalledWith(
      "global",
      appIndexURL,
      appVocabURL
    );
  });

  it("displays error if it fails to load resourceBundles", () => {
    mockResourceBundleLoaderHook(mockedResourceBundleLoaderHook, {
      error: new Error(),
    });
    const { asFragment, getByTestId } = render(
      <AppConfig appIndexURL={appIndexURL} appVocabURL={appVocabURL}>
        test
      </AppConfig>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_ERROR)).toBeDefined();
  });

  it("displays loading while resourceBundles are loading", () => {
    mockResourceBundleLoaderHook(mockedResourceBundleLoaderHook, {
      data: null,
    });
    const { asFragment, getByTestId } = render(
      <AppConfig appIndexURL={appIndexURL} appVocabURL={appVocabURL}>
        test
      </AppConfig>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_LOADING)).toBeDefined();
  });
});
