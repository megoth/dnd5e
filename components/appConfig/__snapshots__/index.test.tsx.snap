// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppConfig displays error if it fails to load resourceBundles 1`] = `
<DocumentFragment>
  <div
    data-testid="error"
  >
    <h1
      data-testid="error-title"
    />
  </div>
</DocumentFragment>
`;

exports[`AppConfig displays loading while resourceBundles are loading 1`] = `
<DocumentFragment>
  <div
    data-testid="loading"
  >
    Loading...
  </div>
</DocumentFragment>
`;

exports[`AppConfig loads resources and renders 1`] = `
<DocumentFragment>
  test
</DocumentFragment>
`;
